### PR TITLE
Refactor some funcs in NSGA-III

### PR DESCRIPTION
**Pull request by nabenabe0928**
## Motivation

This PR aims to simplify the implementation of a function defined in NSGA-III.

## Description of the changes

- Use vectorization better in `_filter_inf`
- Use vectorization better in `generate_default_reference_point`

<details>
<summary>Verification Code of _filter_inf</summary>

```python
import numpy as np


_COEF = 3


def original(objective_matrix: np.ndarray) -> np.ndarray:
    mask_posinf = np.isposinf(objective_matrix)
    mask_neginf = np.isneginf(objective_matrix)

    # Replace +-inf with nan temporary to get max and min.
    objective_matrix[mask_posinf + mask_neginf] = np.nan
    nadir_point = np.nanmax(objective_matrix, axis=0)
    ideal_point = np.nanmin(objective_matrix, axis=0)
    interval = nadir_point - ideal_point

    rows_posinf, cols_posinf = np.where(mask_posinf)
    objective_matrix[rows_posinf, cols_posinf] = (
        nadir_point[cols_posinf] + _COEF * interval[cols_posinf]
    )
    rows_neginf, cols_neginf = np.where(mask_neginf)
    objective_matrix[rows_neginf, cols_neginf] = (
        ideal_point[cols_neginf] - _COEF * interval[cols_neginf]
    )

    return objective_matrix


def this_pr(objective_matrix: np.ndarray) -> np.ndarray:
    objective_matrix_with_nan = np.where(np.isfinite(objective_matrix), objective_matrix, np.nan)
    max_objectives = np.nanmax(objective_matrix_with_nan, axis=0)
    min_objectives = np.nanmin(objective_matrix_with_nan, axis=0)
    margins = _COEF * (max_objectives - min_objectives)
    return np.clip(objective_matrix, min_objectives - margins, max_objectives + margins)


if __name__ == "__main__":
    rng = np.random.RandomState(42)
    for i in range(100):
        print(f"Pass Test #{i}")
        a = np.random.random((100, 3))
        a[a > 0.9] = np.inf
        a[a < 0.1] = -np.inf
        ground_truth = original(a)
        ans = this_pr(a)
        assert np.allclose(ground_truth, ans)
```

</details>

<details>
<summary>Verification Code of generate_default_reference_point</summary>

```python
from itertools import combinations_with_replacement
import math

import numpy as np


def _multi_choose(n: int, k: int) -> int:
    return math.comb(n + k - 1, k)


def original(
    n_objectives: int, dividing_parameter: int = 3
) -> np.ndarray:
    """Generates default reference points which are `uniformly` spread on a hyperplane."""
    reference_points = np.zeros(
        (
            _multi_choose(n_objectives, dividing_parameter),
            n_objectives,
        )
    )
    for i, comb in enumerate(
        combinations_with_replacement(range(n_objectives), dividing_parameter)
    ):
        for j in comb:
            reference_points[i, j] += 1.0
    return reference_points


def this_pr(
    n_objectives: int, dividing_parameter: int = 3
) -> np.ndarray:
    """Generates default reference points which are `uniformly` spread on a hyperplane."""
    indices = np.array(
        list(combinations_with_replacement(range(n_objectives), dividing_parameter))
    )
    row_indices = np.repeat(np.arange(len(indices)), dividing_parameter)
    col_indices = indices.flatten()
    reference_points = np.zeros((len(indices), n_objectives), dtype=float)
    np.add.at(reference_points, (row_indices, col_indices), 1.0)
    return reference_points


if __name__ == "__main__":
    for i in range(3, 7):
        print(original(i), this_pr(i))
        assert np.allclose(original(i), this_pr(i))

```

</details>

### CODE REVIEW

File: optuna/samplers/_nsgaiii/_elite_population_selection_strategy.py [hunk 1]
```diff
@@ -102,55 +101,30 @@ def __call__(self, study: Study, population: list[FrozenTrial]) -> list[FrozenTr
...
+    reference_points = np.zeros((len(indices), n_objectives), dtype=float)
+    np.add.at(reference_points, (row_indices, col_indices), 1.0)
     return reference_points
 
 
 def _filter_inf(population: list[FrozenTrial]) -> np.ndarray:
-    # Collect all objective values.
-    n_objectives = len(population[0].values)
-    objective_matrix = np.zeros((len(population), n_objectives))
+    objective_matrix = np.zeros((len(population), len(population[0].values)))
```
Comment (nabenabe0928) [inline optuna/samplers/_nsgaiii/_elite_population_selection_strategy.py:119]: I am not sure if I can assume that `population[0].values` is not None, but I followed the original code.

**Comment by c-bata** [comment]
@gen740 @HideakiImamura Could you review this PR?

**Review by gen740 — APPROVED** [review]
LGTM!

File: optuna/samplers/_nsgaiii/_elite_population_selection_strategy.py (hunk 1 above)
Comment (nabenabe0928) [reply optuna/samplers/_nsgaiii/_elite_population_selection_strategy.py]: I confirmed that population is a list of completed trials.
- [generation_to_population only includes completed trials](https://github.com/optuna/optuna/blob/master/optuna/samplers/_nsgaiii/_sampler.py#L219-L220)
- [population is created from `generation_to_population`](https://github.com/optuna/optuna/blob/master/optuna/samplers/_nsgaiii/_sampler.py#L227)

**Review by HideakiImamura — APPROVED** [review]
Sorry for the late. LGTM.

**Pull request merged**